### PR TITLE
Add Circuit Breakers panel concept (like in Lagom)

### PR DESCRIPTION
**Pull request by ihostage**
Concept (and implementation of course 😄 ) inspired by Lagom Framework.
I'm sure it can be very useful not only for Lagom users but also for Play Framework and Akka Platform users.
Telemetry inspired by Akka Projection.

Refs https://github.com/akka/akka/issues/16617

### CODE REVIEW

File: akka-actor/src/main/scala/akka/pattern/internal/CircuitBreakersPanelInternal.scala [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      breaker: CircuitBreaker,
+      telemetry: CircuitBreakerTelemetry,
+      failureFn: Try[_] => Boolean)
+
+  private class CircuitBreakerConfig(val _config: Config) {
+    val config: Config = _config.getConfig("akka.circuit-breaker")
+    val default: Config = config.getConfig("default")
+  }
+
+  private val breakers = new ConcurrentHashMap[String, Option[CircuitBreakerHolder]]
```
Comment (ignasi35) [inline akka-actor/src/main/scala/akka/pattern/internal/CircuitBreakersPanelInternal.scala:34]: I think the Lagom impl is leaky https://github.com/lagom/lagom/issues/3304

**Review by patriknw — COMMENTED** [review]
Telemetry for CircuitBreakers is kind of essential so I like the suggestion, but we might need some more iterations on the API...

**Comment by ihostage** [comment]
Hi, @patriknw! Sorry for long silence. I had found a time and reworked this draft by your comments. Can you do a yet review?
Summary of changes:
* `CircuitBreakersPanel` implemented as extension and renamed to `CircuitBreakersRegistry`
* `CircuitBreakerTelemetry` is moved from registry to `CircuitBreaker` directly.

File: akka-actor/src/main/scala/akka/pattern/CircuitBreakersRegistry.scala [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+   * Is used by Akka to instantiate the Extension identified by this ExtensionId,
+   * internal use only.
+   */
+  override def createExtension(system: ExtendedActorSystem): CircuitBreakersRegistry =
+    new CircuitBreakersRegistry(system)
+
+  /**
+   * Returns the canonical ExtensionId for this Extension
+   */
+  override def lookup: ExtensionId[_ <: Extension] = CircuitBreakersRegistry
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/pattern/CircuitBreakersRegistry.scala:30]: two more needed for it to work well from Java:
```
  override def get(system: ActorSystem): CircuitBreakersRegistry = super.get(system)

  override def get(system: ClassicActorSystemProvider): CircuitBreakersRegistry = super.get(system)
```

File: akka-actor/src/main/scala/akka/pattern/internal/CircuitBreakerTelemetry.scala [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        List(classOf[String] -> breakerId, classOf[ExtendedActorSystem] -> system))
+      .get
+  }
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object CircuitBreakerNoopTelemetry extends CircuitBreakerTelemetry {
+  override def onOpen(): Unit = {}
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/pattern/internal/CircuitBreakerTelemetry.scala:113]: for empty methods returning Unit we typically use `= ()`

**Review by patriknw — COMMENTED** [review]
this is looking great

File: akka-actor/src/main/resources/reference.conf [hunk 4]
```diff
@@ -1307,4 +1307,33 @@ akka {
     #//#coordinated-shutdown-phases
   }
 
+  #//#circuit-breaker-default
+  # Circuit breakers. A child configuration section with the same
```
[the remark this answers is not in the document]
Comment (ihostage) [reply akka-actor/src/main/resources/reference.conf:1311]: Thank you, Johan! 👍

File: akka-actor/src/main/scala/akka/pattern/CircuitBreakersRegistry.scala (hunk 2 above)
Comment (ihostage) [reply akka-actor/src/main/scala/akka/pattern/CircuitBreakersRegistry.scala:30]: Done ✅

File: akka-actor/src/main/scala/akka/pattern/internal/CircuitBreakerTelemetry.scala (hunk 3 above)
Comment (ihostage) [reply akka-actor/src/main/scala/akka/pattern/internal/CircuitBreakerTelemetry.scala:113]: Done ✅

**Comment by ihostage** [comment]
Hi, @patriknw! 👋 
Thank you for your comments! 👍 
I added one commit with:
* Add configuration for all circuit breaker params
* Add tests for identified circuit breaker

File: akka-testkit/src/test/resources/reference.conf [hunk 5]
```diff
@@ -31,4 +31,15 @@ akka {
...
+  circuit-breaker {
+    identified {
+      max-failures = 1
+      call-timeout = 100 ms
+      reset-timeout = 200 ms
+      exception-allowlist = [
+        "akka.pattern.CircuitBreakerSpec$AllowException"
+      ]
+    }
+  }
```
Comment (patriknw) [inline akka-testkit/src/test/resources/reference.conf:44]: I would prefer to include this in `CircuitBreakerSpec` instead. It's too hidden here.

```
class CircuitBreakerSpec extends AkkaSpec("""
  akka.circuit-breaker {
     identified {
       max-failures = 1
       call-timeout = 100 ms
       reset-timeout = 200 ms
       exception-allowlist = [
         "akka.pattern.CircuitBreakerSpec$AllowException"
       ]
     }
   }
   """) {
```

**Review by patriknw — COMMENTED** [review]
looking good, just a suggestion for the test

File: akka-testkit/src/test/resources/reference.conf (hunk 5 above)
Comment (ihostage) [reply akka-testkit/src/test/resources/reference.conf]: Done ✅

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by ihostage** [comment]
Thank you, Patrik @patriknw! 🙏 
@johanandren Johan, WDYT?

**Review by johanandren — APPROVED** [review]
LGTM, nice work @ihostage !

**Comment by nicoga97** [comment]
Hi @ihostage, sorry to bother you with this old PR, but I don't find documentation about how to use this circuit breaker metrics they are exposed by Kamon? or how it work, thanks

**Comment by ihostage** [comment]
Hi @nicoga97! I'm not sure that Kamon already supports that. But supporting that is very trivial. 
1. Implement trait [CircuitBreakerTelemetry](https://github.com/akka/akka/blob/3dbcf1c8ea0af031476ca0b50f52bea1d1e5f88e/akka-actor/src/main/scala/akka/pattern/internal/CircuitBreakerTelemetry.scala#L13-L24) like as if I did for Lagom (see [CircuitBreakerMetricsProviderImpl](https://github.com/kamon-io/Kamon/blob/a8bef44040c20047de700bc09f608bbdfdea033b/instrumentation/kamon-lagom/src/main/scala/kamon/instrumentation/lagom/CircuitBreakerMetricsProviderImpl.scala#L32-L78) in Kamon Lagom module)
2. Add implementation by next config 
```HOCON
akka.circuit-breaker.telemetry.implementations += com.example.MyMetrics
```

**Comment by nicoga97** [comment]
@ihostage Thanks for the fast response! Great seems easy!

**Pull request merged**